applications/skyepass.md

```diff
@@ -0,0 +1,251 @@
+# Open Grant Proposal
+
+> This document is referenced in the terms and conditions and therefore needs to contain all the required information. Don't remove any of the mandatory parts presented in bold letters or as headlines! See the [Open Grants Program Process](https://github.com/w3f/Open-Grants-Program/blob/master/README_2.md) on how to submit a proposal.
+
+* **Project Name:** SkyePass
+* **Team Name:** SkyeKiwi Team
+* **Payment Address**: 0xa5E4E1BB29eE2D16B07545CCf565868aE34F92a2
+
+*The above combination of your GitHub account submitting the application and payment address will be your unique identifier during the program. Please keep them safe.*
+
+## Project Overview :page_facing_up: 
+
+SkyePass is a decentralized and customizable identity management software. On the surface, it is a decentralized, open source and modern password manager. 
+
+### Product Details
+
+As a long-term password manager software user myself, I have been really frustrated of the services like LastPass, 1Password for either lack of functionalities or the idea of storing ones entire digital identity on their corporate servers. Existing open-source solutions are too technically complicated to use. 
+
+At the very basis of it, a password manager is no more than an encrypted database, an APP and a browser extension to interact with the database. 
+
+Therefore, our team create a new password manger software that has pretty and intuitive UI/UX, fully decentralized (i.e. our team own no backend servers) and hackable by providing an open API for people to develop extensions with. 
+
+Users who signup will first create a blockchain wallet and have the mnemonic (and a master password) as their sole identity credentials (pretty standard blockchain wallet stuff). Later, each database instance is called a `vault` (standard name for all password managers) and they are light-weight file based databases ([lowDB](https://github.com/typicode/lowdb) seems to be a great choice). User can be given options to choose the encryption behavior of their database. By default, the vault will be split into some pieces with a Shamir's secret sharing mechanism. 
+
+For instance, for a simplest sharing schema, when the vault is created to be shared with 2 other family members, the vault will be split into 4 parts (we call them `horcrux`, for those who do not know [the Harry Potter reference](http://harrypotter.shoutwiki.com/wiki/Horcrux#:~:text=A%20Horcrux%20is%20a%20powerful,one%20is%20to%20true%20immortality.) ) with a minimum quorum of 2 to be decrypted. One piece will be sent to IPFS without encryption, the other 3 pieces will be encrypted by each member's public key and be sent to IPFS. An NFT will be minted for the owner. The ID of the NFT will be the `vault ID` and the NFT's URI will be a metadata piece that only the owner can change as exampled below: 
```

Comment:
If it's split into four parts and everyone has access to the unencrypted part, then threshold = 2 means any user can unlock the secret (own share + public IPFS share), no? Unless I'm missing something, this doesn't sound like the intended design.

Reply:
Not sure if I understand it right. For "user", you mean members of a vault whose shares are encrypted with their public keys, right? If so, that's the intended design for a simplest sharing, like a vault I wanna share with my family members. The capacity of creating more complicated schemas (for teams/business) is reserved but will not be our main focus yet. After delivering these two milestones, two huge extensible capacities will be more focused: more complex sharing schema and extensions/integrated apps/wallets.

Reply:
Sorry for any misunderstandings. Allow me to clarify (this is how I understand your protocol, correct me if I'm wrong please):
1. A principal user (A) wishes to store some password, so they choose 2 trusted users B & C, who could help them with password recovery. There are a total of 3 (real) users.
2. If I understand your design, the secret is split into 4 parts: 3 are encrypted with A, B, C's public keys and one part is always sent unencrypted to IPFS, so it's accessible to all parties involved (and also outside parties)
3. We set a threshold of T = 2, because we want any two users to be able to recover the secret phrase.

Did I get it correctly until now?
If so, then this is where the design has a flow: right now any **single** user is able to recover the secret (because they know how to decrypt their own share + they have access to the public share). I don't fully understand the need for IPFS unencrypted share here...If you want 2/3 recovery, you should just send the parts to A, B, C and that's it - any two of them are able to cooperate to recover the key. Otherwise, you might as well do 1/3 recovery (which doesn't make much sense by definition).
I'm not an expert in cryptography so I might be missing some parts of the protocol - if so please rectify my mistakes.

Reply:
I can see your confusion now. Your understanding of the schema is technically correct but I guess you understand the intention behind it off. Not sure if you are familiar with the team vault/family shared vault by 1Password. We are trying to build something like that without a centralized server. 

The logic that lead to such schema is a case when you are trying to share a small end-to-end encrypted database to some trustees and **invite them to contribute to the same database** and of course, **without a centralized server**. A most naive way would be send them the whole database encrypted with their public keys, upload to a file system and let them know the URL. However, when any of the members updates the database, it will be messy to ensure the integrity of the database and very hard to enforce a `write` and `read` privilege management, it can falls into some kinda `git` like file-diff algo at best, blockchain-like consensus algo at worst. To stay away from these rabbit holes and implement easy and effective access control, Shamir is the algo designed to solve sharing the database and blockchain smart contract is the perfect tool for access control and preserve change history of the database. Changes are sequential on a blockchain. So the metadata should be neat and tidy. But it is not a perfect world, and members can corrupt the metadata. So we designed this system of having a public unencrypted piece. When a user fails to decrypt with the public piece, they can easily come to know that the metadata is corrupted. Actions can be take, either roll back to the most recent working version or force whoever corrupted the metadata to fix it.

I guess the confusion you had was on whether users actually need to read and write the database. 

Later, we realized that, the flexibility that Shamir offers has a wider application and is a prefect fit for a password manager. For instance, a business team, with some help of their private servers, can make more complicated and customized sharing schema and access control. For individuals to leave a legacy (i.e. two private pieces of the owner, 1 private piece to the lawyer, 2 private pieces to 2 trustees, 0 for public pieces and 3 pieces to decrypt). Shamir is very flexible, so that a vault owner can upgrade the schema with minimum efforts. Given that the blockchain preserves all the metadata history, but the linked IPFS data can be deleted, so kicking someone off is also doable. 

Also, as I said earlier, SkyePass is able to support more complicated sharing schema other than sharing with trusted family members when everyone has access to read and to write. We can produce more sharing options, scripts for business to setup their private backup servers, customized IPFS data preservation policy etc. We want to explore this part later because: 1. I will be honest and say, we got at most 20% of this part figured out. Identity and access management can get arbitrarily complicated. But we do not want to invest too much of our time thinking of all kinds of encryption schema yet, because we want to build solutions to problems. We plan to talk and understand how teams want their vaults managed before we build solutions for them. So far, I think it's enough for us to have a simple schema and be confident that more complicated schemas will be compatible with all other pieces of SkyePass. 2. For the early stage, these ideas will really confuse people. For me personally, nothing confuses me more than a solution to a problem that I don't understand. And that's exactly the case for IAM systems. 

Sorry for the long reply. Two things in summary: 1. users can read and edit the vaults, by design. They are not there for backup or recovery but as collaborators to the vault. 2. We introduce Shamir and this seemingly odd public unencrypted piece to reserve the capacity for more advanced features in the future, with compatibilities and extensibility design in mind.